### PR TITLE
Audio: add NeMo FastConformer ASR adapter

### PR DESCRIPTION
**Pull request by mohammadaaftabv**
## Summary

Adds a NeMo ASR adapter for the shared `ASRStage` / `ASRAdapter` contract and migrates Curator's in-repository NeMo ASR consumers to that adapter-based path.

The resulting ownership split is:

- `ASRStage` owns file decoding, mono conversion, resampling, language filtering, batching, skip handling, output assembly, resources, and lifecycle routing.
- `NeMoASRAdapter` owns NeMo checkpoint download, worker-local model load/unload, optional FastConformer local-attention conversion, optional RNNT CUDA-graph decoder configuration, batched transcription, and NeMo output normalization.

## Changes

- Add `NeMoASRAdapter` behind the shared ASR adapter protocol.
- Keep model-specific options in `adapter_kwargs` rather than forcing them onto every ASR adapter.
- Migrate the FLEURS benchmark, YAML, README, and notebook to `ASRStage` + `NeMoASRAdapter`.
- Add a minimal shared-runner FastConformer tutorial using `README.md` + `pipeline.yaml`.
- Update the shared YAML runner so `pipeline` is the default Hydra config name.
- Keep Qwen-Omni model revision adapter-owned while preserving its existing pipeline behavior.
- Move TorchCodec into the audio-common dependency path and select its CPU wheel for audio installs while retaining the Torch 2.11-compatible version override already present on `main`.
- Update focused unit, configuration, lazy-import, stage, tagging, and real-GPU smoke coverage.

## Intentional migration

This PR removes the legacy `nemo_curator.stages.audio.inference.asr.asr_nemo.InferenceAsrNemoStage` module and migrates all in-repository consumers to the shared stage/adapter path. It does not add a compatibility re-export; downstream users of the legacy import must move to:

```python
ASRStage(
    adapter_target="nemo_curator.models.asr.nemo_asr.NeMoASRAdapter",
    model_id="nvidia/stt_en_fastconformer_ctc_large",
)
```

## Review feedback addressed

- Removed redundant adapter type/value checks.
- Moved `torch` and OmegaConf imports to module scope while retaining the optional NeMo lazy import.
- Consolidated the GPU smoke test into `tests/models/asr/test_nemo_asr.py`.
- Made `--config-name pipeline` optional for shared-runner tutorials and added CLI coverage.
- Removed the exact direct TorchCodec pin and documented the ABI-compatible shared override.

## Validation after rebasing onto current `main`

- 126 focused CPU tests passed; the marked real-GPU smoke test was deselected in this CPU test command.
- Ruff lint passed.
- CI-equivalent pre-commit checks passed.
- `uv lock --check` passed with uv 0.12.2.
- `git diff --check` passed.

### CODE REVIEW

File: nemo_curator/models/asr/nemo_asr.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    verbose: bool = False
+    enable_local_attention: bool = False
+    local_attention_context_size: tuple[int, int] = (128, 128)
+    use_cuda_graph_decoder: bool | None = None
+    refresh_cache: bool = False
+    strict: bool = True
+    _model: Any = field(default=None, init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        if not self.model_id:
```
Comment (sarahyurick) [inline nemo_curator/models/asr/nemo_asr.py]: Some of these checks are overkill IMO. Like this and the check for whether `enable_local_attention` (and others) is a boolean.

File: nemo_curator/models/asr/nemo_asr.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+    def load_model(self, *, num_gpus: int) -> None:
+        """Load one worker-local model on the device requested by ``ASRStage``."""
+        if self._model is not None:
+            return
+        if isinstance(num_gpus, bool) or not isinstance(num_gpus, Integral) or num_gpus not in {0, 1}:
+            msg = f"NeMoASRAdapter requires num_gpus to be 0 or 1, got {num_gpus!r}"
+            raise ValueError(msg)
+
+        import torch
```
Comment (sarahyurick) [inline nemo_curator/models/asr/nemo_asr.py]: Let's use top-level imports for most of these.

File: tests/models/asr/test_nemo_asr_gpu.py
Comment (sarahyurick) [inline tests/models/asr/test_nemo_asr_gpu.py:1]: Let's move all the tests to `test_nemo_asr`.

File: tutorials/audio/nemo_fastconformer/README.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+For CPU-only use, install `audio_cpu` instead.
+
+## Run the bundled smoke input
+
+The repository already contains a two-row manifest and two short OPUS files:
+
+```bash
+python nemo_curator/config/run.py \
+  --config-path ../../tutorials/audio/nemo_fastconformer \
+  --config-name pipeline \
```
Comment (sarahyurick) [inline tutorials/audio/nemo_fastconformer/README.md]: Can `config-name` be optional?

File: pyproject.toml [hunk 4]
```diff
@@ -112,6 +112,7 @@ audio_common = [
     "nemo_toolkit[asr]>=2.7.2; platform_machine == 'x86_64' and platform_system != 'Darwin'",
     "soundfile>=0.12.0",
     "torchaudio",
+    "torchcodec==0.11.1; platform_machine == 'x86_64' and platform_system != 'Darwin'",
```
Comment (sarahyurick) [inline pyproject.toml]: Why the pin?

File: tutorials/audio/fleurs/fleurs_tutorial.ipynb [hunk 5]
```diff
@@ -526,137 +353,66 @@
...
+     "iopub.status.busy": "2026-08-05T16:42:33.373675Z",
+     "iopub.status.idle": "2026-08-05T16:42:36.731825Z",
+     "shell.execute_reply": "2026-08-05T16:42:36.731067Z"
     }
    },
    "outputs": [],
    "source": [
-    "configure_loguru()\n",
-    "ray_client.stop()"
+    "configure_loguru()\nray_client.stop()"
```
Comment (sarahyurick) [inline tutorials/audio/fleurs/fleurs_tutorial.ipynb]: Just wondering why do we need `configure_loguru` here?

File: pyproject.toml [hunk 6]
```diff
@@ -385,6 +385,11 @@ name = "vllm-cu129"
 url = "https://wheels.vllm.ai/0.22.0/cu129"
 explicit = true
 
+[[tool.uv.index]]
+name = "pytorch-cpu"
```
Comment (sarahyurick) [inline pyproject.toml]: I was also wondering why it needs this?

**Comment by mohammadaaftabv** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Curator/pull/2254/commits/9e38391b5eda6c9ba183552563ef8d98cca126a9

File: pyproject.toml [hunk 7]
```diff
@@ -399,8 +399,7 @@ torchvision = [
     { index = "pypi", marker = "sys_platform == 'darwin' or (platform_machine != 'x86_64' and platform_machine != 'aarch64')" },
 ]
 torchcodec = [
-    { index = "pytorch", marker = "platform_machine == 'x86_64' and sys_platform != 'darwin'" },
-    { index = "pypi", marker = "platform_machine != 'x86_64' or sys_platform == 'darwin'" },
+    { index = "pypi", extra = "audio-common" },
```
Comment (sarahyurick) [inline pyproject.toml:404]: Sorry just want to double check the logic here. This is compatible with installing `audio_cpu`, `audio_cuda12`, and `all`?

Comment (mohammadaaftabv) [reply pyproject.toml:404]: yes for the repository’s documented uv installation paths:

1. audio_cpu includes audio_common.
2. audio_cuda12 includes audio_common.
3. all includes audio_cuda12, which includes audio_common.

Therefore, all three activate the extra = "audio-common" PyPI source rule.

The override constrains TorchCodec to ~=0.11.0; the lockfile resolves 0.11.1 from PyPI.
Locked dry runs for all three extras successfully selected torchcodec==0.11.1.
The actual CUDA environment has Torch 2.11.0+cu129 with TorchCodec 0.11.1+cpu, and both audio tutorials passed.

This source rule affects only TorchCodec. Existing Torch/Torchaudio index selection is unchanged.
uv officially supports selecting dependency sources by optional extra in this way. [uv dependency-source documentation](https://github.com/astral-sh/uv/blob/main/docs/concepts/projects/dependencies.md)

**Review by sarahyurick — APPROVED** [review]

File: nemo_curator/models/asr/nemo_asr.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+from typing import Any
+
+import numpy as np
+import torch
+from omegaconf import open_dict
+
+from nemo_curator.models.asr.base import ASRResult
+
+_DEFAULT_FASTCONFORMER_CTC_MODEL = "nvidia/stt_en_fastconformer_ctc_large"
+_DEFAULT_SAMPLE_RATE = 16_000
```
Comment (virajkarandikar) [inline nemo_curator/models/asr/nemo_asr.py]: all of these can be moved inside `NeMoASRAdapter`

**Review by virajkarandikar — APPROVED** [review]
minor comment.

**Comment by mohammadaaftabv** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Curator/pull/2254/commits/524189eb2cf75fe23b7c78868074d38dbe1a079d

**Comment by mohammadaaftabv** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Curator/pull/2254/commits/9a7ef2c6ccfbbee49f7e8941ac0bc6aee410ad5a

**Comment by mohammadaaftabv** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Curator/pull/2254/commits/16a0c811104f84fbe9105c3b03ac52285146f1c4
